<a href="https://colab.research.google.com/github/Kayumov094/Iinitial/blob/main/birinchi_clean_data_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

![Imgur](https://i.imgur.com/h4PPC71.png)
#11-modul. Maʼlumotlarga ishlov berish. Maʼlumotlarni tayyorlash
# 26. Modul boʻyicha yakuniy amaliy vazifa

**Bu vazifalarni bajargach, maʼlumotlarni yuklash, tozalash, oʻzgartirish, matnlarni normalizatsiya qilish, qiymatlarni tekshirish, guruhlash va tahlil qilish kabi maʼlumotlarga ishlov berish (data processing) koʻnikmalarini egallaysiz.**

#**Umumiy maʼlumotlar bilan ishlash**

#Dataset haqida qisqacha maʼlumot
Restoran savdosi boʻyicha maʼlumotlar toʻplami 17 534 tranzaksiya uchun maʼlumotlarni oʻz ichiga oladi. Maʼlumotlar yetishmayotgan yoki baʼzilari toʻliq kiritilmagan boʻlishi mumkin. Maʼlumotlar toʻplamida asosiy taomlar, shirinliklar, ichimliklar kabi bir nechta toifadagi savdo tafsilotlari mavjud.


 0.   **Unnamed 0** – indekslarni;
 1.   **Order ID** – buyurtma IDʼsini;
 2.   **Customer ID** – mijoz IDʼsini;
 3.   **Category** – mahsulotlar kategoriyasini;
 4.   **Item** – mahsulot nomini;
 5.   **Price** – mahsulotning narxini;
 6.   **Quantity** – qancha mahsulot sotilgani miqdorini;
 7.   **Order Total** – umumiy narx qancha ekanini;
 8.   **Order Date** – buyurtma berilgan sanani (vaqt);
 9.   **Payment Method** – toʻlov usuli turini saqlaydi.

#1) Quyida berilgan datasetni yuklab olib, ustun nomlarini koʻring va ularni maʼno jihatdan tushunarli shaklga keltiring. `Unnamed: 0` ustunini tashlab yuboring!

In [ ]:
import pandas as pd
import numpy as np

url = "https://raw.githubusercontent.com/alishermutalov/praktikum-datasets/refs/heads/praktikum/restaurant_sales.csv"
df = pd.read_csv(url)

df.head()

Unnamed: 0    Order ID Customer ID     Category             Item  Price  \
0           0  ORD_705844    CUST_092          NaN     side salad#?    3.0   
1           1  ORD_338528    CUST_021  Side Dishes  Mashed Potatoes    4.0   
2           2  ORD_443849    CUST_029  Main Dishes  Grilled Chicken   15.0   
3           3  ORD_630508    CUST_075       Drinks              NaN   -1.0   
4           4  ORD_648269    CUST_031  Main Dishes    Pasta Alfredo   12.0   

   Quantity  Order Total  Order Date  Payment Method  
0       1.0          3.0  2023-12-21     Credit Card  
1       3.0         12.0  2023-05-19  Digital Wallet  
2       4.0         60.0  2023-09-27     Credit Card  
3       2.0          5.0  2022-08-09     Credit Card  
4       4.0         48.0  2022-05-15            Cash

In [ ]:
df.drop(["Unnamed: 0"], axis=1, inplace=True) # Unnamed:0 ustunini tashlab yubordim

In [ ]:
df

Order ID Customer ID     Category             Item  Price  Quantity  \
0      ORD_705844    CUST_092          NaN     side salad#?    3.0       1.0   
1      ORD_338528    CUST_021  Side Dishes  Mashed Potatoes    4.0       3.0   
2      ORD_443849    CUST_029  Main Dishes  Grilled Chicken   15.0       4.0   
3      ORD_630508    CUST_075       Drinks              NaN   -1.0       2.0   
4      ORD_648269    CUST_031  Main Dishes    Pasta Alfredo   12.0       4.0   
...           ...         ...          ...              ...    ...       ...   
17529  ORD_320102    CUST_021       Drinks              NaN   -1.0       4.0   
17530  ORD_974128    CUST_069     Desserts        Ice Cream    5.0       3.0   
17531  ORD_108324    CUST_050     Desserts        Ice Cream    5.0       4.0   
17532  ORD_612647    CUST_073  Side Dishes  Mashed Potatoes    4.0       2.0   
17533  ORD_680707    CUST_006     Desserts              NaN    7.0       4.0   

       Order Total  Order Date  Payment Method  
0              3.0  2023-12-21     Credit Card  
1             12.0  2023-05-19  Digital Wallet  
2             60.0  2023-09-27     Credit Card  
3              5.0  2022-08-09     Credit Card  
4             48.0  2022-05-15            Cash  
...            ...         ...             ...  
17529          4.0  2023-12-23            Cash  
17530         15.0  2023-12-18            Cash  
17531         20.0  2022-05-20  Digital Wallet  
17532          8.0  2022-01-27  Digital Wallet  
17533         28.0  2023-01-07             NaN  

[17534 rows x 9 columns]

#2) Maʼlumotlarning umumiy holatini (info(), describe()) tekshiring.

In [ ]:
df.info(), df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17534 entries, 0 to 17533
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Order ID        17534 non-null  object 
 1   Customer ID     17534 non-null  object 
 2   Category        17498 non-null  object 
 3   Item            15776 non-null  object 
 4   Price           17534 non-null  float64
 5   Quantity        17104 non-null  float64
 6   Order Total     17104 non-null  float64
 7   Order Date      17534 non-null  object 
 8   Payment Method  16452 non-null  object 
dtypes: float64(3), object(6)
memory usage: 1.2+ MB


(None,
               Price      Quantity   Order Total
 count  17534.000000  17104.000000  17104.000000
 mean       6.207312      3.014149     19.914494
 std        4.993782      1.414598     18.732549
 min       -1.000000      1.000000      1.000000
 25%        3.000000      2.000000      7.500000
 50%        5.000000      3.000000     15.000000
 75%        7.000000      4.000000     25.000000
 max       20.000000      5.000000    100.000000)

In [ ]:
df

Order ID Customer ID     Category             Item  Price  Quantity  \
0      ORD_705844    CUST_092          NaN     side salad#?    3.0       1.0   
1      ORD_338528    CUST_021  Side Dishes  Mashed Potatoes    4.0       3.0   
2      ORD_443849    CUST_029  Main Dishes  Grilled Chicken   15.0       4.0   
3      ORD_630508    CUST_075       Drinks              NaN   -1.0       2.0   
4      ORD_648269    CUST_031  Main Dishes    Pasta Alfredo   12.0       4.0   
...           ...         ...          ...              ...    ...       ...   
17529  ORD_320102    CUST_021       Drinks              NaN   -1.0       4.0   
17530  ORD_974128    CUST_069     Desserts        Ice Cream    5.0       3.0   
17531  ORD_108324    CUST_050     Desserts        Ice Cream    5.0       4.0   
17532  ORD_612647    CUST_073  Side Dishes  Mashed Potatoes    4.0       2.0   
17533  ORD_680707    CUST_006     Desserts              NaN    7.0       4.0   

       Order Total  Order Date  Payment Method  
0              3.0  2023-12-21     Credit Card  
1             12.0  2023-05-19  Digital Wallet  
2             60.0  2023-09-27     Credit Card  
3              5.0  2022-08-09     Credit Card  
4             48.0  2022-05-15            Cash  
...            ...         ...             ...  
17529          4.0  2023-12-23            Cash  
17530         15.0  2023-12-18            Cash  
17531         20.0  2022-05-20  Digital Wallet  
17532          8.0  2022-01-27  Digital Wallet  
17533         28.0  2023-01-07             NaN  

[17534 rows x 9 columns]

#3) Har bir ustundagi boʻsh qiymatlarni aniqlang va ularni toʻldirish strategiyasini ishlab chiqing.

In [ ]:
df

Order ID Customer ID     Category             Item  Price  Quantity  \
0      ORD_705844    CUST_092          NaN     side salad#?    3.0       1.0   
1      ORD_338528    CUST_021  Side Dishes  Mashed Potatoes    4.0       3.0   
2      ORD_443849    CUST_029  Main Dishes  Grilled Chicken   15.0       4.0   
3      ORD_630508    CUST_075       Drinks              NaN   -1.0       2.0   
4      ORD_648269    CUST_031  Main Dishes    Pasta Alfredo   12.0       4.0   
...           ...         ...          ...              ...    ...       ...   
17529  ORD_320102    CUST_021       Drinks              NaN   -1.0       4.0   
17530  ORD_974128    CUST_069     Desserts        Ice Cream    5.0       3.0   
17531  ORD_108324    CUST_050     Desserts        Ice Cream    5.0       4.0   
17532  ORD_612647    CUST_073  Side Dishes  Mashed Potatoes    4.0       2.0   
17533  ORD_680707    CUST_006     Desserts              NaN    7.0       4.0   

       Order Total  Order Date  Payment Method  
0              3.0  2023-12-21     Credit Card  
1             12.0  2023-05-19  Digital Wallet  
2             60.0  2023-09-27     Credit Card  
3              5.0  2022-08-09     Credit Card  
4             48.0  2022-05-15            Cash  
...            ...         ...             ...  
17529          4.0  2023-12-23            Cash  
17530         15.0  2023-12-18            Cash  
17531         20.0  2022-05-20  Digital Wallet  
17532          8.0  2022-01-27  Digital Wallet  
17533         28.0  2023-01-07             NaN  

[17534 rows x 9 columns]

In [ ]:
df.columns, df.index # ustunlar va qator index haqida malumotlarni oldim


(Index(['Order ID', 'Customer ID', 'Category', 'Item', 'Price', 'Quantity',
        'Order Total', 'Order Date', 'Payment Method'],
       dtype='object'),
 RangeIndex(start=0, stop=17534, step=1))

# keyingi qadam priceni minuslarini olib tashlash

In [ ]:
df["Price"] = df["Price"].replace({-1:1}) # replace() dan foyalanib -1 ni 1 ga almashtirdim
df

Order ID Customer ID     Category             Item  Price  Quantity  \
0      ORD_705844    CUST_092          NaN     side salad#?    3.0       1.0   
1      ORD_338528    CUST_021  Side Dishes  Mashed Potatoes    4.0       3.0   
2      ORD_443849    CUST_029  Main Dishes  Grilled Chicken   15.0       4.0   
3      ORD_630508    CUST_075       Drinks              NaN    1.0       2.0   
4      ORD_648269    CUST_031  Main Dishes    Pasta Alfredo   12.0       4.0   
...           ...         ...          ...              ...    ...       ...   
17529  ORD_320102    CUST_021       Drinks              NaN    1.0       4.0   
17530  ORD_974128    CUST_069     Desserts        Ice Cream    5.0       3.0   
17531  ORD_108324    CUST_050     Desserts        Ice Cream    5.0       4.0   
17532  ORD_612647    CUST_073  Side Dishes  Mashed Potatoes    4.0       2.0   
17533  ORD_680707    CUST_006     Desserts              NaN    7.0       4.0   

       Order Total  Order Date  Payment Method  
0              3.0  2023-12-21     Credit Card  
1             12.0  2023-05-19  Digital Wallet  
2             60.0  2023-09-27     Credit Card  
3              5.0  2022-08-09     Credit Card  
4             48.0  2022-05-15            Cash  
...            ...         ...             ...  
17529          4.0  2023-12-23            Cash  
17530         15.0  2023-12-18            Cash  
17531         20.0  2022-05-20  Digital Wallet  
17532          8.0  2022-01-27  Digital Wallet  
17533         28.0  2023-01-07             NaN  

[17534 rows x 9 columns]

Keyingi qadam kategoriya bilan ishlash

In [ ]:
df["Category"].isnull().value_counts() # Categoty da 36 qiymatda null mavjud

Category
False    17498
True        36
Name: count, dtype: int64

In [ ]:
category_nul = df[['Category']].isnull().sum() # series qiymat  36 ta null
df.Category.value_counts(dropna=False) # Category ustunini unikal qiymatlarni chiqarib oldim

Category
Main Dishes    3544
Starters       3517
Desserts       3493
Drinks         3479
Side Dishes    3465
NaN              36
Name: count, dtype: int64

In [ ]:
df.groupby("Category")["Price"].value_counts()

Category     Price
Desserts     6.0      1336
             5.0       996
             7.0       502
             4.0       479
             1.0       180
Drinks       1.0      1203
             2.5      1140
             3.0      1136
Main Dishes  12.0     1058
             15.0      864
             20.0      596
             18.0      443
             14.0      403
             1.0       180
Side Dishes  4.0      1247
             3.0      1038
             5.0      1021
             1.0       159
Starters     5.0      1198
             4.0       944
             7.0       494
             8.0       464
             10.0      247
             1.0       170
Name: count, dtype: int64

In [ ]:
df.groupby("Price")["Item"].value_counts(dropna=False)

Price  Item       
1.0    NaN            942
       Water (0.5)    902
       WATER           19
       water           16
       WATER#          10
                     ... 
20.0   STEAK           20
       STEAK#           8
       steak            7
       STEAK?           2
       Steak?           2
Name: count, Length: 190, dtype: int64

In [ ]:
df.loc[(df['Price'] == 1.0) & (df['Category'].isnull()), 'Category'] = 'Drinks'#narxlarni guruhladim va unga categoriyalarni ham qo'shdik.
# NaN qiymatlarnini qo'shib qiymatlarini hioblab chiqdik
# Demak bizda 1.0 dollarda Nan Categotiya 6 ta
#2.5 dollarda Nan Categotiya 4 ta
#3.0 dollarda Nan Categotiya 6 ta
#4.0 dollarda Nan Categotiya 3 ta
#5.0 dollarda Nan Categotiya 5 ta
#6.0 dollarda Nan Categotiya 1 ta
#7.0 dollarda Nan Categotiya 2 ta
#8.0 dollarda Nan Categotiya 2 ta
#12.0 dollarda Nan Categotiya 1 ta
#15.0 dollarda Nan Categotiya 2 ta
#18.0 dollarda Nan Categotiya 2 ta
#20.0 dollarda Nan Categotiya 1 ta ekanligini aniqladik ura

Category da NaN qiymatlarni har bir narxdagi category ning modasiga qarab ffill qildim

In [ ]:
df.loc[(df['Price'] == 2.5) & (df['Category'].isnull()), 'Category'] = 'Drinks'
df.loc[(df['Price'] == 3.0) & (df['Category'].isnull()), 'Category'] = 'Drinks'
df.loc[(df['Price'] == 4.0) & (df['Category'].isnull()), 'Category'] = 'Side Dishes'
df.loc[(df['Price'] == 5.0) & (df['Category'].isnull()), 'Category'] = 'Starters'
df.loc[(df['Price'] == 6.0) & (df['Category'].isnull()), 'Category'] = 'Desserts'
df.loc[(df['Price'] == 7.0) & (df['Category'].isnull()), 'Category'] = 'Desserts'
df.loc[(df['Price'] == 8.0) & (df['Category'].isnull()), 'Category'] = 'Starters'
df.loc[(df['Price'] == 12.0) & (df['Category'].isnull()), 'Category'] = 'Main Dishes'
df.loc[(df['Price'] == 15.0) & (df['Category'].isnull()), 'Category'] = 'Main Dishes'
df.loc[(df['Price'] == 18.0) & (df['Category'].isnull()), 'Category'] = 'Main Dishes'

In [ ]:
df['Price'].value_counts() # bizda narxlar to'liq to'drilgan. Category ni toldirishni narxlar orqali boramiz
# Price ning qiymatlarni sanab oldik

Price
5.0     3221
4.0     2673
3.0     2180
1.0     1898
6.0     1337
2.5     1144
12.0    1059
7.0      998
15.0     866
20.0     597
8.0      466
18.0     445
14.0     403
10.0     247
Name: count, dtype: int64

In [ ]:
unikal_narx = df['Price'].unique() # price ning unikal qiymatlarini chiqarib oldim
unikal_narx

array([ 3. ,  4. , 15. ,  1. , 12. , 18. ,  6. , 14. ,  5. ,  2.5,  7. ,
        8. , 10. , 20. ])

In [ ]:
unikal_narx.sort() # unikal narxni tartiblab oldim

In [ ]:
unikal_narx # Unikla narxlarni chiqaraman, biroq manda -1 narx bor ekan. shuni plusga aylantiraman.

array([ 1. ,  2.5,  3. ,  4. ,  5. ,  6. ,  7. ,  8. , 10. , 12. , 14. ,
       15. , 18. , 20. ])

In [ ]:
df["Price"] = df["Price"].replace(-1, 1) # -1 narxlarni o'rniga 1 narxlarga almashtirdik
unikal_narx = df['Price'].unique() # price ning unikal qiymatlarini chiqarib oldim
unikal_narx.sort()
unikal_narx

array([ 1. ,  2.5,  3. ,  4. ,  5. ,  6. ,  7. ,  8. , 10. , 12. , 14. ,
       15. , 18. , 20. ])

In [ ]:
df["Price"].value_counts().sort_values()

Price
10.0     247
14.0     403
18.0     445
8.0      466
20.0     597
15.0     866
7.0      998
12.0    1059
2.5     1144
6.0     1337
1.0     1898
3.0     2180
4.0     2673
5.0     3221
Name: count, dtype: int64

In [ ]:
df.groupby("Price")["Category"].value_counts(dropna = False)

Price  Category   
1.0    Drinks         1209
       Desserts        180
       Main Dishes     180
       Starters        170
       Side Dishes     159
2.5    Drinks         1144
3.0    Drinks         1142
       Side Dishes    1038
4.0    Side Dishes    1250
       Starters        944
       Desserts        479
5.0    Starters       1204
       Side Dishes    1021
       Desserts        996
6.0    Desserts       1337
7.0    Desserts        504
       Starters        494
8.0    Starters        466
10.0   Starters        247
12.0   Main Dishes    1059
14.0   Main Dishes     403
15.0   Main Dishes     866
18.0   Main Dishes     445
20.0   Main Dishes     596
       NaN               1
Name: count, dtype: int64

Price ustuni bo'yicha categorylarni modasini aniqlab oldim. sababi Nan qiymatlarni to'ldirish uchun

In [ ]:
moda_bir = df[(df['Price'] == 1.0)]['Category'].mode()[0]            # price 1.0 bolgan mahsulotlarning modasini topib oldim
moda_ikki_yarim = df[(df['Price'] == 2.5)]['Category'].mode()[0]     # price 2.5 bolgan mahsulotlarning modasini topib oldim
moda_uch = df[(df['Price'] == 3.0)]['Category'].mode()[0]            # price 3.0 bolgan mahsulotlarning modasini topib oldim
moda_tort = df[(df['Price'] == 4.0)]['Category'].mode()[0]           # price 4.0 bolgan mahsulotlarning modasini topib oldim
moda_besh = df[(df['Price'] == 5.0)]['Category'].mode()[0]           # price 5.0 bolgan mahsulotlarning modasini topib oldim
moda_olti = df[(df['Price'] == 6.0)]['Category'].mode()[0]           # price 6.0 bolgan mahsulotlarning modasini topib oldim
moda_yetti = df[(df['Price'] == 7.0)]['Category'].mode()[0]          # price 7.0 bolgan mahsulotlarning modasini topib oldim
moda_sakkiz = df[(df['Price'] == 8.0)]['Category'].mode()[0]         # price 8.0 bolgan mahsulotlarning modasini topib oldim
moda_on_ikki = df[(df['Price'] == 12.0)]['Category'].mode()[0]       # price 12.0 bolgan mahsulotlarning modasini topib oldim
moda_on_besh = df[(df['Price'] == 15.0)]['Category'].mode()[0]       # price 15.0 bolgan mahsulotlarning modasini topib oldim
moda_on_sakkiz = df[(df['Price'] == 18.0)]['Category'].mode()[0]     # price 18.0 bolgan mahsulotlarning modasini topib oldim
moda_yigirma = df[(df['Price'] == 20.0)]['Category'].mode()[0]       # price 20.0 bolgan mahsulotlarning modasini topib oldim


In [ ]:
df.loc[(df['Price'] == 1) & (df['Category'].isnull()), 'Category'] = 'Drinks'
# bu yerda aytilyaptiki narxi 1 ga teng va categoriyasi null bolgan categoriya ustunidagi qiymatni "Drinks" bilan to'ldir dedim
# sababi 1 dollarlik narxda eng ko'p categoriya "Drinks " ekan.
#qolgan null qiymatlarni shu strategiya orqali to'ldirib chiqaman.

In [ ]:
df.loc[(df['Price'] == 2.5) & (df['Category'].isnull()), 'Category'] = 'Drinks'
df.loc[(df['Price'] == 3.0) & (df['Category'].isnull()), 'Category'] = 'Drinks'
df.loc[(df['Price'] == 4.0) & (df['Category'].isnull()), 'Category'] = 'Side Dishes'
df.loc[(df['Price'] == 5.0) & (df['Category'].isnull()), 'Category'] = 'Starters'
df.loc[(df['Price'] == 6.0) & (df['Category'].isnull()), 'Category'] = 'Desserts'
df.loc[(df['Price'] == 7.0) & (df['Category'].isnull()), 'Category'] = 'Desserts'
df.loc[(df['Price'] == 8.0) & (df['Category'].isnull()), 'Category'] = 'Starters'
df.loc[(df['Price'] == 12.0) & (df['Category'].isnull()), 'Category'] = 'Main Dishes'
df.loc[(df['Price'] == 15.0) & (df['Category'].isnull()), 'Category'] = 'Main Dishes'
df.loc[(df['Price'] == 18.0) & (df['Category'].isnull()), 'Category'] = 'Main Dishes'
df.loc[(df['Price'] == 20.0) & (df['Category'].isnull()), 'Category'] = 'Main Dishes'

Category da null qiymatlar qolmadi. keyin bosqich "Item" nul qiymatlar bilan toldirish. Oldin Itemni tartibga keltirib olaman

In [ ]:
df['Item'].isnull().sum() # demak Item ustunida 1758 ta null qiymatlar mavjud

np.int64(1758)

In [ ]:
# Item ustununi tartibga keltirib olaman
df['Item'] = df['Item'].str.strip()
df['Item'] = df['Item'].str.replace(r'[^\w\s]', '', regex=True)
df['Item'] = df['Item'].str.lower()
df['Item'] = df['Item'].replace({'water#': 'water', 'water?': 'water', "water 05":"water"})
df['Item'] = df['Item'].replace({'lemonade 10': 'lemonade'})
df['Category'] = df['Category'].str.lower()
df['Payment Method'] = df['Payment Method'].str.lower()

In [ ]:
df.Item.unique()

array(['side salad', 'mashed potatoes', 'grilled chicken', nan,
       'pasta alfredo', 'salmon', 'garlic bread', 'vegetarian platter',
       'brownie', 'sweet potato fries', 'lemonade', 'orange juice',
       'chocolate cake', 'coca cola', 'grilled vegetables',
       'cheese fries', 'beef chili', 'chicken melt', 'fruit salad',
       'french fries', 'ice cream', 'cheesecake', 'iced tea', 'water',
       'nachos grande', 'onion rings', 'steak'], dtype=object)

In [ ]:
df

Order ID Customer ID     Category             Item  Price  Quantity  \
0      ORD_705844    CUST_092       drinks       side salad    3.0       1.0   
1      ORD_338528    CUST_021  side dishes  mashed potatoes    4.0       3.0   
2      ORD_443849    CUST_029  main dishes  grilled chicken   15.0       4.0   
3      ORD_630508    CUST_075       drinks              NaN    1.0       2.0   
4      ORD_648269    CUST_031  main dishes    pasta alfredo   12.0       4.0   
...           ...         ...          ...              ...    ...       ...   
17529  ORD_320102    CUST_021       drinks              NaN    1.0       4.0   
17530  ORD_974128    CUST_069     desserts        ice cream    5.0       3.0   
17531  ORD_108324    CUST_050     desserts        ice cream    5.0       4.0   
17532  ORD_612647    CUST_073  side dishes  mashed potatoes    4.0       2.0   
17533  ORD_680707    CUST_006     desserts              NaN    7.0       4.0   

       Order Total  Order Date  Payment Method  
0              3.0  2023-12-21     credit card  
1             12.0  2023-05-19  digital wallet  
2             60.0  2023-09-27     credit card  
3              5.0  2022-08-09     credit card  
4             48.0  2022-05-15            cash  
...            ...         ...             ...  
17529          4.0  2023-12-23            cash  
17530         15.0  2023-12-18            cash  
17531         20.0  2022-05-20  digital wallet  
17532          8.0  2022-01-27  digital wallet  
17533         28.0  2023-01-07             NaN  

[17534 rows x 9 columns]

In [ ]:
df.groupby("Price")[["Category","Item"]].value_counts(dropna = False).head(100)  # Price ni category va Item bo'yicha guruhlab NaN qiymatlarini chiqarib oldim

Price  Category     Item              
1.0    drinks       water                 956
                    NaN                   253
       desserts     NaN                   180
       main dishes  NaN                   180
       starters     NaN                   170
       side dishes  NaN                   159
2.5    drinks       coca cola             756
                    iced tea              328
                    NaN                    60
3.0    side dishes  side salad            975
       drinks       orange juice          591
                    lemonade              479
                    NaN                    69
       side dishes  NaN                    63
       drinks       side salad              3
4.0    starters     french fries          895
       side dishes  mashed potatoes       799
       desserts     fruit salad           449
       side dishes  garlic bread          399
                    NaN                    50
       starters     NaN                    49
       desserts     NaN                    30
       side dishes  french fries            2
5.0    desserts     ice cream             934
       starters     cheese fries          686
       side dishes  grilled vegetables    576
       starters     sweet potato fries    471
       side dishes  onion rings           373
                    NaN                    72
       desserts     NaN                    62
       starters     NaN                    43
                    grilled vegetables      2
                    ice cream               2
6.0    desserts     chocolate cake        798
                    brownie               469
                    NaN                    70
7.0    desserts     cheesecake            485
       starters     beef chili            468
                    NaN                    26
       desserts     NaN                    17
                    beef chili              2
8.0    starters     chicken melt          442
                    NaN                    24
10.0   starters     nachos grande         238
                    NaN                     9
12.0   main dishes  pasta alfredo         998
                    NaN                    61
14.0   main dishes  vegetarian platter    382
                    NaN                    21
15.0   main dishes  grilled chicken       822
                    NaN                    44
18.0   main dishes  salmon                422
                    NaN                    23
20.0   main dishes  steak                 574
                    NaN                    23
Name: count, dtype: int64

In [ ]:
mahsulot_turi = df.Item.value_counts(dropna=False) # mahhsulot turlarning unikal qiymatlarni chiqarib oldim demak 1758 ta NAN bor /max/

In [ ]:
Item_nullar = df[(df['Item'].isna())] #Item dagi nullar hammasini chiqarib oldim

In [ ]:
Item_nullar.groupby("Price")["Category"].value_counts(dropna = False)

Price  Category   
1.0    drinks         253
       desserts       180
       main dishes    180
       starters       170
       side dishes    159
2.5    drinks          60
3.0    drinks          69
       side dishes     63
4.0    side dishes     50
       starters        49
       desserts        30
5.0    side dishes     72
       desserts        62
       starters        43
6.0    desserts        70
7.0    starters        26
       desserts        17
8.0    starters        24
10.0   starters         9
12.0   main dishes     61
14.0   main dishes     21
15.0   main dishes     44
18.0   main dishes     23
20.0   main dishes     23
Name: count, dtype: int64

Item ustunini toldirish uchun funksiyani yaratib oldim


In [ ]:
# Narxlar ro'yxati
prices = [1.0, 2.5, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 10.0, 12.0, 14.0, 15.0, 18.0, 20.0]

# Har bir narx uchun Item NaN qiymatlarni to'ldirish
for price in prices:
    # Narxga mos DataFrame yaratish
    category_mode = df[df['Price'] == price].groupby('Category')['Item'].agg(lambda x: x.mode())

    # NaN qiymatlarni moda bilan to'ldirish
    def fill_item_with_mode(row):
        if pd.isna(row['Item']) and row['Price'] == price:
            return category_mode.get(row['Category'], row['Item'])
        return row['Item']

    # DataFrame'da NaN qiymatlarni to'ldirish
    df['Item'] = df.apply(fill_item_with_mode, axis=1)

<ipython-input-316-06eff0217803>:11: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if pd.isna(row['Item']) and row['Price'] == price:
<ipython-input-316-06eff0217803>:11: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if pd.isna(row['Item']) and row['Price'] == price:
<ipython-input-316-06eff0217803>:11: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if pd.isna(row['Item']) and row['Price'] == price:
<ipython-input-316-06eff0217803>:11: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use

In [ ]:
df['Item'].isna().value_counts()

Item
False    17534
Name: count, dtype: int64

Item ustuni ham toliq toldirishni yakunladim

Quantity UStunida 430 ta nan qiymatlar bor ekan

In [ ]:
quantity_nan = df[df['Quantity'].isna()]
quantity_nan

Order ID Customer ID     Category   Item  Price  Quantity  \
75     ORD_367023    CUST_055  side dishes     []    1.0       NaN   
118    ORD_627458    CUST_058  side dishes     []    1.0       NaN   
131    ORD_467157    CUST_099       drinks  water    1.0       NaN   
240    ORD_467033    CUST_032  side dishes     []    1.0       NaN   
253    ORD_722632    CUST_027     desserts     []    1.0       NaN   
...           ...         ...          ...    ...    ...       ...   
17410  ORD_882909    CUST_023       drinks  water    1.0       NaN   
17451  ORD_529422    CUST_068       drinks  water    1.0       NaN   
17473  ORD_342327    CUST_096     desserts     []    1.0       NaN   
17497  ORD_205432    CUST_090       drinks  water    1.0       NaN   
17506  ORD_563553    CUST_044  main dishes     []    1.0       NaN   

       Order Total  Order Date  Payment Method  
75             NaN  2022-05-29             NaN  
118            NaN  2023-08-24  digital wallet  
131            NaN  2023-11-30             NaN  
240            NaN  2023-03-04  digital wallet  
253            NaN  2022-12-11     credit card  
...            ...         ...             ...  
17410          NaN  2023-07-11     credit card  
17451          NaN  2022-07-17     credit card  
17473          NaN  2022-12-29            cash  
17497          NaN  2022-03-03            cash  
17506          NaN  2022-07-20  digital wallet  

[430 rows x 9 columns]

In [ ]:
quantity_nan.groupby("Price")["Category"].value_counts(dropna = False)

Price  Category   
1.0    desserts       93
       drinks         93
       main dishes    89
       starters       81
       side dishes    74
Name: count, dtype: int64

Quantity ustuni Nan qiymatlar kamligi uchun umumiy modaga qarab to'ldirib oldim. Sababi qolgan qiymatlarning standart ogishi uncha katta emas


In [ ]:
df["Quantity"] = df["Quantity"].fillna(df['Quantity'].mode()[0])
df["Quantity"].isna().value_counts()

Quantity
False    17534
Name: count, dtype: int64

In [ ]:
df.info() # info bilan tanishdim kyingi qadam Order Total ustuni bilan ishlash

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17534 entries, 0 to 17533
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Order ID        17534 non-null  object 
 1   Customer ID     17534 non-null  object 
 2   Category        17534 non-null  object 
 3   Item            17534 non-null  object 
 4   Price           17534 non-null  float64
 5   Quantity        17534 non-null  float64
 6   Order Total     17104 non-null  float64
 7   Order Date      17534 non-null  object 
 8   Payment Method  16452 non-null  object 
dtypes: float64(3), object(6)
memory usage: 1.2+ MB


:Quantity ustuni ham moda() yordamida to'ldirildi. endi keyingi ustunga o'tamiz

In [ ]:
df["Order Total"].isna().value_counts() # demak 430 ta NaN qauymat bor

Order Total
False    17104
True       430
Name: count, dtype: int64

In [ ]:
df_raqam = df.select_dtypes(include='number')
df_raqam.corr() # xulosa Order total Price bilan eng katta bog'liqlikka ega

Price  Quantity  Order Total
Price        1.000000 -0.028046     0.764124
Quantity    -0.028046  1.000000     0.502331
Order Total  0.764124  0.502331     1.000000

In [ ]:
Order_nan = df[df["Order Total"].isna()]
Order_nan.groupby("Price")["Category"].value_counts(dropna = False)

Price  Category   
1.0    desserts       93
       drinks         93
       main dishes    89
       starters       81
       side dishes    74
Name: count, dtype: int64

In [ ]:
Order_nan[Order_nan["Price"]>1] # demal bizda Order Total ning barcha narxlari 1 ga teng ekan
# xulosani df ning narxi bir ga teng bolganlarni analiz qilish orqali amalga oshiraman

Empty DataFrame
Columns: [Order ID, Customer ID, Category, Item, Price, Quantity, Order Total, Order Date, Payment Method]
Index: []

In [ ]:
df[df['Price']<=1] .groupby("Price")["Order Total"].value_counts(dropna = False)
# df.groupby("Price")[["Category", "Order Total"]].value_counts()


Price  Order Total
1.0    NaN            430
       5.0            248
       3.0            230
       4.0            229
       1.0            211
       2.0            199
       12.0            35
       20.0            33
       10.0            26
       25.0            25
       15.0            24
       6.0             18
       8.0             16
       16.0            15
       30.0            12
       9.0             11
       24.0            11
       18.0            10
       2.5              9
       14.0             9
       28.0             9
       60.0             8
       80.0             8
       7.0              6
       35.0             6
       36.0             6
       48.0             6
       40.0             5
       45.0             5
       32.0             4
       42.0             4
       100.0            4
       7.5              3
       12.5             3
       21.0             3
       70.0             3
       72.0             3
       90.0             3
       50.0             2
       54.0             2
       56.0             2
       75.0             2
Name: count, dtype: int64

In [ ]:
df[(df['Price']<=1.0) & (df["Order Total"]<=1.0)].groupby("Price")[["Order Total", "Category"]].value_counts(dropna = False)

Price  Order Total  Category
1.0    1.0          drinks      211
Name: count, dtype: int64

Yuqoridagi natijadan xulosa qilish mumkinki narxi 1.0 bolgan buyurtmalar categoriyasi "drinks" ekan. kichik xulosa sifatida drinksdagi nan qiymatlarni order total=1.0 boladi.  

In [ ]:
order_nan_drinks = Order_nan[(Order_nan["Price"] < 2.5) & (Order_nan["Category"] == 'drinks')]
Order_nan["Order Total"].fillna(1.0, inplace=True)

<ipython-input-328-72ef5ed22138>:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  Order_nan["Order Total"].fillna(1.0, inplace=True)
<ipython-input-328-72ef5ed22138>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Order_nan["Order Total"].fillna(1.0, inplace=True)


In [ ]:
df.loc[Order_nan.index, "Order Total"] = Order_nan["Order Total"] # ajratib olinga order_nan ustunlari orginal Order Total ustunlariga almashtirildi

Order Totak ustuni ham to'liq toldirildi

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17534 entries, 0 to 17533
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Order ID        17534 non-null  object 
 1   Customer ID     17534 non-null  object 
 2   Category        17534 non-null  object 
 3   Item            17534 non-null  object 
 4   Price           17534 non-null  float64
 5   Quantity        17534 non-null  float64
 6   Order Total     17534 non-null  float64
 7   Order Date      17534 non-null  object 
 8   Payment Method  16452 non-null  object 
dtypes: float64(3), object(6)
memory usage: 1.2+ MB


In [ ]:
df["Payment Method"].unique() # unikal qiymatlarni chiqarib oldim

array(['credit card', 'digital wallet', 'cash', nan], dtype=object)

In [ ]:
payment_notna = df[df['Payment Method'].notna()] # payment method ustunida nan siz qiymatlar chiqarib oldim endi ularni tahlil qilaman
payment_notna.groupby("Price")["Category"].value_counts(dropna = False)

Price  Category   
1.0    drinks         1099
       desserts        122
       starters        121
       main dishes     118
       side dishes     110
2.5    drinks         1093
3.0    drinks         1069
       side dishes     991
4.0    side dishes    1197
       starters        902
       desserts        455
5.0    starters       1156
       side dishes     973
       desserts        952
6.0    desserts       1270
7.0    desserts        481
       starters        468
8.0    starters        445
10.0   starters        233
12.0   main dishes     998
14.0   main dishes     383
15.0   main dishes     827
18.0   main dishes     414
20.0   main dishes     575
Name: count, dtype: int64

In [ ]:
payment_notna.groupby("Order Total")["Payment Method"].value_counts(dropna = False) # guruhlab oldik

Order Total  Payment Method
1.0          digital wallet    175
             credit card       167
             cash              150
2.0          cash               65
             credit card        61
                              ... 
90.0         credit card        25
             digital wallet     25
100.0        cash               48
             credit card        39
             digital wallet     36
Name: count, Length: 123, dtype: int64

In [ ]:
foiz_notna = payment_notna.groupby("Order Total")["Payment Method"].value_counts(normalize=True, dropna=False) * 100
foiz_notna

Order Total  Payment Method
1.0          digital wallet    35.569106
             credit card       33.943089
             cash              30.487805
2.0          cash              34.946237
             credit card       32.795699
                                 ...    
90.0         credit card       30.120482
             digital wallet    30.120482
100.0        cash              39.024390
             credit card       31.707317
             digital wallet    29.268293
Name: proportion, Length: 123, dtype: float64

In [ ]:
df

Order ID Customer ID     Category             Item  Price  Quantity  \
0      ORD_705844    CUST_092       drinks       side salad    3.0       1.0   
1      ORD_338528    CUST_021  side dishes  mashed potatoes    4.0       3.0   
2      ORD_443849    CUST_029  main dishes  grilled chicken   15.0       4.0   
3      ORD_630508    CUST_075       drinks            water    1.0       2.0   
4      ORD_648269    CUST_031  main dishes    pasta alfredo   12.0       4.0   
...           ...         ...          ...              ...    ...       ...   
17529  ORD_320102    CUST_021       drinks            water    1.0       4.0   
17530  ORD_974128    CUST_069     desserts        ice cream    5.0       3.0   
17531  ORD_108324    CUST_050     desserts        ice cream    5.0       4.0   
17532  ORD_612647    CUST_073  side dishes  mashed potatoes    4.0       2.0   
17533  ORD_680707    CUST_006     desserts       cheesecake    7.0       4.0   

       Order Total  Order Date  Payment Method  
0              3.0  2023-12-21     credit card  
1             12.0  2023-05-19  digital wallet  
2             60.0  2023-09-27     credit card  
3              5.0  2022-08-09     credit card  
4             48.0  2022-05-15            cash  
...            ...         ...             ...  
17529          4.0  2023-12-23            cash  
17530         15.0  2023-12-18            cash  
17531         20.0  2022-05-20  digital wallet  
17532          8.0  2022-01-27  digital wallet  
17533         28.0  2023-01-07             NaN  

[17534 rows x 9 columns]

In [ ]:
payment_nan = df[df['Payment Method'].isnull()]
payment_nan.groupby("Price")["Category"].value_counts(dropna = False)

Price  Category   
1.0    drinks         110
       main dishes     62
       desserts        58
       side dishes     49
       starters        49
2.5    drinks          51
3.0    drinks          73
       side dishes     47
4.0    side dishes     53
       starters        42
       desserts        24
5.0    side dishes     48
       starters        48
       desserts        44
6.0    desserts        67
7.0    starters        26
       desserts        23
8.0    starters        21
10.0   starters        14
12.0   main dishes     61
14.0   main dishes     20
15.0   main dishes     39
18.0   main dishes     31
20.0   main dishes     22
Name: count, dtype: int64

In [ ]:
print(df['Item'].apply(type).value_counts())
df['Item'] = df['Item'].apply(lambda x: str(x) if isinstance(x, np.ndarray) else x)
print(df['Item'].apply(type).value_counts())

Item
<class 'str'>              16845
<class 'numpy.ndarray'>      689
Name: count, dtype: int64
Item
<class 'str'>    17534
Name: count, dtype: int64


In [ ]:
df["Payment Method"] = df["Payment Method"].fillna(df["Payment Method"].mode()[0])
df["Payment Method"].isna().value_counts()

Payment Method
False    17534
Name: count, dtype: int64

In [ ]:
df.info() # jadval to'liq to'ldirildi

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17534 entries, 0 to 17533
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Order ID        17534 non-null  object 
 1   Customer ID     17534 non-null  object 
 2   Category        17534 non-null  object 
 3   Item            17534 non-null  object 
 4   Price           17534 non-null  float64
 5   Quantity        17534 non-null  float64
 6   Order Total     17534 non-null  float64
 7   Order Date      17534 non-null  object 
 8   Payment Method  17534 non-null  object 
dtypes: float64(3), object(6)
memory usage: 1.2+ MB


#4) Har bir ustundagi takroriy qiymatlarni toping va ularga qanday oʻzgartirish kiritish kerakligini hal qiling.

In [ ]:
# javobni shu yerda yozing
df['Customer ID'].duplicated().value_counts() # Customer ID da

Customer ID
True     17434
False      100
Name: count, dtype: int64

In [ ]:
df[df['Customer ID'].duplicated()].value_counts()

Order ID    Customer ID  Category     Item             Price  Quantity  Order Total  Order Date  Payment Method
ORD_999939  CUST_033     side dishes  mashed potatoes  4.0    3.0       12.0         2022-11-26  digital wallet    1
ORD_100067  CUST_037     desserts     ice cream        5.0    1.0       5.0          2023-04-23  cash              1
ORD_100118  CUST_056     starters     cheese fries     5.0    5.0       25.0         2023-09-04  credit card       1
ORD_100127  CUST_058     main dishes  steak            20.0   5.0       100.0        2023-07-31  cash              1
ORD_100145  CUST_015     starters     french fries     4.0    2.0       8.0          2022-09-28  cash              1
                                                                                                                  ..
ORD_100626  CUST_087     drinks       lemonade         3.0    2.0       6.0          2022-12-16  credit card       1
ORD_100577  CUST_063     desserts     ice cream        5.0    1.0       5.0          2022-05-21  digital wallet    1
ORD_100569  CUST_002     starters     french fries     4.0    5.0       20.0         2022-08-19  digital wallet    1
ORD_100552  CUST_093     side dishes  side salad       3.0    3.0       9.0          2023-12-22  digital wallet    1
ORD_100549  CUST_064     drinks       orange juice     3.0    1.0       3.0          2022-09-21  credit card       1
Name: count, Length: 17434, dtype: int64

#5) Maʼlumotlar toʻplamidagi noto‘g‘ri maʼlumotlar (masalan, son turishi kerak bo‘lgan ustunda matn bo‘lsa) mavjudligini tekshiring va to‘g‘rilang.

dataframe tozalandi


In [ ]:
df

Order ID Customer ID     Category             Item  Price  Quantity  \
0      ORD_705844    CUST_092       drinks       side salad    3.0       1.0   
1      ORD_338528    CUST_021  side dishes  mashed potatoes    4.0       3.0   
2      ORD_443849    CUST_029  main dishes  grilled chicken   15.0       4.0   
3      ORD_630508    CUST_075       drinks            water    1.0       2.0   
4      ORD_648269    CUST_031  main dishes    pasta alfredo   12.0       4.0   
...           ...         ...          ...              ...    ...       ...   
17529  ORD_320102    CUST_021       drinks            water    1.0       4.0   
17530  ORD_974128    CUST_069     desserts        ice cream    5.0       3.0   
17531  ORD_108324    CUST_050     desserts        ice cream    5.0       4.0   
17532  ORD_612647    CUST_073  side dishes  mashed potatoes    4.0       2.0   
17533  ORD_680707    CUST_006     desserts       cheesecake    7.0       4.0   

       Order Total  Order Date  Payment Method  
0              3.0  2023-12-21     credit card  
1             12.0  2023-05-19  digital wallet  
2             60.0  2023-09-27     credit card  
3              5.0  2022-08-09     credit card  
4             48.0  2022-05-15            cash  
...            ...         ...             ...  
17529          4.0  2023-12-23            cash  
17530         15.0  2023-12-18            cash  
17531         20.0  2022-05-20  digital wallet  
17532          8.0  2022-01-27  digital wallet  
17533         28.0  2023-01-07     credit card  

[17534 rows x 9 columns]

#6) Maʼlumotlar toʻplamidagi **katta/kichik harf** muammolari boʻlgan joylarni toping va bir xil formatga keltiring.

In [ ]:
#javobni shu yerda yozing
df["Order ID"] = df["Order ID"].str.strip()
df["Customer ID"] = df["Customer ID"].str.strip()
df["Category"] = df["Category"].str.strip()
df["Item"] = df["Item"].str.strip()
df["Payment Method"] = df["Payment Method"].str.strip()

#7) Maʼlumotlar turini oʻzgartirish.
 * Ustunlarning maʼlumot turlarini tekshiring. Notoʻgʻri turlarga ega boʻlgan ustunlarni toʻgʻri formatga oʻtkazing.
 * `order_date` yoki unga oʻxshash ustunni datetime formatiga oʻtkazing.

In [ ]:
#javobni shu yerda yozing
df['Order Date'] = pd.to_datetime(df['Order Date'])
type(df['Order Date'])

pandas.core.series.Series

#8) Mahsulot nomlarida noto‘g‘ri yoki nomaʼlum belgilar (?, NaN, #) mavjudligini tekshiring va ularni tuzating.

boshida tozalab olgan edim

#**Maʼlumotlarni filtrlash va qayta ishlash**

#9) Manfiy narx yoki 0 ga teng narx qiymatlarini toping va ularni qanday oʻzgartirish kerakligini hal qiling.

yuqrida manfiy sonlarni va nol qoymatlarni mantiqiy tartibda sozlab olganman

#10) Boʻsh toifalar (kategoriya) mavjudligini tekshiring va toʻgʻri kategoriya berish strategiyasini ishlab chiqing.

jadval 100% amntiqiy ketma-ketlikda to'ldirdim

In [ ]:
df.loc[df["Item"] == "[]", "Item"] = "selfi"

#11) Bir xil, lekin turlicha yozilgan qiymatlar (Coffee, cofee, COFEE) mavjudligini tekshiring va tuzating.

dataframe ichidagi xar xil yozilganlar bir xillikka keltirildi

#12) `Customer ID` yoki boshqa identifikatorlarning takrorlanishini tekshiring.

In [ ]:
#javobni shu yerda yozing
Customer ID tekshirganda 90 % takrorlangan deb hisoblayapti. lekin ularga tegishli boshqa ko'rsatkichlar turlicha
nomlanishda oxirgi 033 yoki 037 sonlarni takrorlangan deb berdi. shu narsaga tushuna olmadim

#13) Mahsulotning ortiqcha taʼriflarini olib tashlang (Water (0.5), Water (1.0), Water → faqat Water qoldirish).

dataframe ichidagi xar xil yozilganlar bir xillikka keltirildi

#**Qayta ishlangan maʼlumotlar bilan tahlil**

#14) Eng koʻp sotilgan 5 ta mahsulotni aniqlang.

In [ ]:
# Eng ko'p sotilgan 5 ta mahsulot
top_mahsulot = df.groupby("Item")["Quantity"].sum().sort_values(ascending=False).head(5)
print(top_mahsulot)

Item
water            3809.0
pasta alfredo    3290.0
side salad       3125.0
ice cream        3091.0
french fries     2857.0
Name: Quantity, dtype: float64


#15) Har bir mijoz oʻrtacha qancha pul sarflaganini hisoblang.

In [ ]:
# har bir mijoz sarflagan pul miqdori
sarflangan_pullar_1 = df.groupby("Customer ID")["Price"].sum().sort_values(ascending=False)
sarflangan_pullar_2 = df.groupby("Customer ID")["Price"].mean().sort_values(ascending=False)
print(
    f"Har bir mijoz sarflagan pul miqdori: {sarflangan_pullar_1}",
    f"\nHar bir mijoz sarflagan o‘rtacha pul miqdori: {sarflangan_pullar_2}"
)

Har bir mijoz sarflagan pul miqdori: Customer ID
CUST_066    1357.5
CUST_020    1349.0
CUST_089    1324.5
CUST_055    1317.5
CUST_054    1314.5
             ...  
CUST_060     943.0
CUST_095     941.5
CUST_003     938.0
CUST_005     909.0
CUST_085     905.5
Name: Price, Length: 100, dtype: float64 
Har bir mijoz sarflagan o‘rtacha pul miqdori: Customer ID
CUST_020    7.213904
CUST_089    7.198370
CUST_054    7.067204
CUST_080    6.967647
CUST_006    6.905063
              ...   
CUST_069    5.576471
CUST_062    5.550279
CUST_011    5.550000
CUST_090    5.497409
CUST_013    5.442529
Name: Price, Length: 100, dtype: float64


#16) Mijozlarning qaysi kunlarda koʻproq xarid qilgani bo‘yicha tahlil qiling.

In [ ]:
#qaysi kunlarda mijoz ko'proq xarid qilgan
savdo_kunlar = df.groupby("Order Date")["Customer ID"].count().sort_values(ascending=False)
print(savdo_kunlar)

Order Date
2023-11-25    42
2022-03-31    41
2022-04-02    40
2023-08-22    40
2023-08-17    37
              ..
2022-07-07    13
2022-02-15    13
2023-04-22    13
2023-08-27    12
2023-01-16    11
Name: Customer ID, Length: 730, dtype: int64


#17) Narxi oʻrtacha narxdan ikki barobar yuqori boʻlgan mahsulotlarni toping.

In [ ]:
# narxning o'rtachadan ikki marta yuqori qiymati
ortachadan_ikki_marta_yuqorilar = df[df["Price"]> df["Price"].mean()*2]
ortachadan_ikki_marta_yuqorilar

Order ID Customer ID     Category                Item  Price  \
2      ORD_443849    CUST_029  main dishes     grilled chicken   15.0   
5      ORD_381680    CUST_031  main dishes              salmon   18.0   
7      ORD_146656    CUST_077  main dishes     grilled chicken   15.0   
9      ORD_743636    CUST_085  main dishes  vegetarian platter   14.0   
11     ORD_726419    CUST_025  main dishes              salmon   18.0   
...           ...         ...          ...                 ...    ...   
17453  ORD_591820    CUST_082  main dishes              salmon   18.0   
17480  ORD_771570    CUST_093  main dishes              salmon   18.0   
17490  ORD_900549    CUST_047  main dishes              salmon   18.0   
17495  ORD_872487    CUST_017  main dishes               steak   20.0   
17505  ORD_605079    CUST_089  main dishes     grilled chicken   15.0   

       Quantity  Order Total Order Date  Payment Method  
2           4.0         60.0 2023-09-27     credit card  
5           5.0         90.0 2022-07-20  digital wallet  
7           3.0         45.0 2023-02-15            cash  
9           5.0         70.0 2022-08-07     credit card  
11          5.0         90.0 2023-10-30            cash  
...         ...          ...        ...             ...  
17453       3.0         54.0 2023-12-08     credit card  
17480       2.0         36.0 2023-03-24     credit card  
17490       2.0         36.0 2023-04-25  digital wallet  
17495       5.0        100.0 2023-11-01     credit card  
17505       4.0         60.0 2022-12-23            cash  

[2311 rows x 9 columns]

#18) Mahsulotlar kategoriyalari bo‘yicha tahlil qiling va eng ko‘p sotiladigan kategoriyalarni aniqlang.

In [ ]:
#Eng ko'p sotilgan mahsulot category
top_category = df.groupby("Category")["Quantity"].count().sort_values(ascending=False)
top_category

Category
main dishes    3550
starters       3525
desserts       3496
drinks         3495
side dishes    3468
Name: Quantity, dtype: int64

#19) Har bir mijozning oʻrtacha xarid miqdorini (`groupby("customer_id")`) hisoblang.

In [ ]:
#jhar bir mijozning o'rtacha harid miqdori
mijoz_haridi = df.groupby("Customer ID")["Quantity"].mean().sort_values(ascending=False)
print(mijoz_haridi)

Customer ID
CUST_046    3.389610
CUST_032    3.321637
CUST_079    3.305085
CUST_095    3.293333
CUST_042    3.250000
              ...   
CUST_024    2.882353
CUST_073    2.879121
CUST_064    2.864516
CUST_009    2.859649
CUST_083    2.793939
Name: Quantity, Length: 100, dtype: float64


#20) Qanday omillar sotuv hajmiga ta’sir qiladi? Bu bo‘yicha umumiy xulosa chiqaring. Javobni quyida kommentariya sifatida yozing!

"Mahatma Gandi" Restorani uchun tahlil
Ushbu tahlil  2022 yil 1 yanvardan 2023-yil 31-dekabr kunigacha restoran savdosi tranzaksiyalari asosida olib borildi.  
Mavjud boʻyicha ma'lumotlar  17 534 tranzaksiya;
kategiriyalar drinks','side dishes','main dishes','desserts','starters';
26 turdagi mahsulotlar hisoblanadi.

Sotuv hajmiga quyidagi omillar ta'sir ko'rsatadi:
1) mahsulotlar turini optimallashtirish. masalan shu davr mobaynida quyidagi
water            3809.0
pasta alfredo    3290.0
side salad       3125.0
ice cream        3091.0
french fries     2857.0 sotuvi eng yuqori ko'rsatkichni ko'rsatdi.

2) Narxlarni maqbullashtirish:
Har bir mijoz sarflagan pul miqdori: 1357.5 dan 905.5 gachani tashkil qiladi.
Har bir mijoz sarflagan o‘rtacha pul miqdori: 7.213904 dan 5.442529 gachani tashkil etadi.
Bundan tashqari restoranda o'rtachan ikki barobar yuqori mahsulotlar sotuvi 2311 tani yoki jamining 13 % tashkil etadi xolos.
Demak Restoranda mahsulotalar narxi 5 dan 15 gacha tayyorlash ko'proq samara olib kelishi mumkin.

3) Xaridorlarga keshbeklar taqdim etish.
tahlil davomida mijozlarning o'rtacha xaridi 3.389610 dan 2.793939 gachani tashkil etadi. demak aholi mazkur restorandan arzonroq mahsulotlar xarid qilish qi;ishga moyil. Ularga keshbek taqdim etish orqali qimmatroq mahsulotlar savdosini ham oshirish mumkim.
Xulosa "Mahatma Gandi" restoranining sotuv hajmiganing o'zgarishiga mahsulotlarning turlari, mahsulotlarning narxlari va xaridorlarning iqtisodiy imkoniyatlari ta'sir ko'rsatadi.  

Fikrim- amaliy mashg'ulot juda juda foydali bo'ldi. agar xatolarni tekshirib feedback berilsa vapshe daxshat bo'ladi. hamma ustozlarga rahmat. Biz o'rganishda davom etamiz




  

In [ ]:
df_orginal = df.copy()
df_orginal

Order ID Customer ID     Category             Item  Price  Quantity  \
0      ORD_705844    CUST_092       drinks       side salad    3.0       1.0   
1      ORD_338528    CUST_021  side dishes  mashed potatoes    4.0       3.0   
2      ORD_443849    CUST_029  main dishes  grilled chicken   15.0       4.0   
3      ORD_630508    CUST_075       drinks            water    1.0       2.0   
4      ORD_648269    CUST_031  main dishes    pasta alfredo   12.0       4.0   
...           ...         ...          ...              ...    ...       ...   
17529  ORD_320102    CUST_021       drinks            water    1.0       4.0   
17530  ORD_974128    CUST_069     desserts        ice cream    5.0       3.0   
17531  ORD_108324    CUST_050     desserts        ice cream    5.0       4.0   
17532  ORD_612647    CUST_073  side dishes  mashed potatoes    4.0       2.0   
17533  ORD_680707    CUST_006     desserts       cheesecake    7.0       4.0   

       Order Total Order Date  Payment Method  
0              3.0 2023-12-21     credit card  
1             12.0 2023-05-19  digital wallet  
2             60.0 2023-09-27     credit card  
3              5.0 2022-08-09     credit card  
4             48.0 2022-05-15            cash  
...            ...        ...             ...  
17529          4.0 2023-12-23            cash  
17530         15.0 2023-12-18            cash  
17531         20.0 2022-05-20  digital wallet  
17532          8.0 2022-01-27  digital wallet  
17533         28.0 2023-01-07     credit card  

[17534 rows x 9 columns]